src/V3CfgLiveVariables.cpp

```diff
@@ -0,0 +1,245 @@
+//
+// Classical data flow analysis computing live variables input to a CFG
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Ast.h"
+#include "V3Cfg.h"
+
+#include <limits>
+#include <memory>
+#include <unordered_map>
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+template <bool T_Scoped>
+class CfgLiveVariables final : VNVisitorConst {
+    // TYPES
+    using Variable = std::conditional_t<T_Scoped, AstVarScope, AstVar>;
+
+    struct BlockState final {
+        std::unordered_set<Variable*> m_gen;
+        std::unordered_set<Variable*> m_kill;
+        std::unordered_set<Variable*> m_liveIn;
+        std::unordered_set<Variable*> m_liveOut;
+        bool m_isOnWorkList = false;
+        bool m_wasProcessed = false;  // Already processed at least once
+    };
+
+    // STATE
+    const ControlFlowGraph& m_cfg;  // The CFG beign analysed
+    BasicBlockMap<BlockState> m_blockState{m_cfg};  // State for each block
+    BlockState* m_currp = nullptr;  // State of current block being analysed
+    bool m_abort = false;  // Abort analysis - unhandled construct
+
+    // METHODS
+    static Variable* getTarget(const AstNodeVarRef* refp) {
+        // TODO: remove the useless reinterpret_casts when C++17 'if constexpr' actually works
+        if VL_CONSTEXPR_CXX17 (T_Scoped) {
+            return reinterpret_cast<Variable*>(refp->varScopep());
+        } else {
+            return reinterpret_cast<Variable*>(refp->varp());
+        }
+    }
+
+    static AstVar* getAstVar(Variable* varp) {
+        // TODO: remove the useless reinterpret_casts when C++17 'if constexpr' actually works
+        if VL_CONSTEXPR_CXX17 (T_Scoped) {
+            return reinterpret_cast<AstVarScope*>(varp)->varp();
+        } else {
+            return reinterpret_cast<AstVar*>(varp);
+        }
+    }
+
+    // This is to match DFG, but can be extended independently
+    static bool isSupportedPackedDType(const AstNodeDType* dtypep) {
+        dtypep = dtypep->skipRefp();
+        if (const AstBasicDType* const typep = VN_CAST(dtypep, BasicDType)) {
+            return typep->keyword().isIntNumeric();
+        }
+        if (const AstPackArrayDType* const typep = VN_CAST(dtypep, PackArrayDType)) {
+            return isSupportedPackedDType(typep->subDTypep());
+        }
+        if (const AstNodeUOrStructDType* const typep = VN_CAST(dtypep, NodeUOrStructDType)) {
+            return typep->packed();
+        }
+        return false;
+    }
+
+    // Check and return if variable is incompatible
+    bool incompatible(Variable* varp) {
+        if (!isSupportedPackedDType(varp->dtypep())) return true;
+        AstVar* const astVarp = getAstVar(varp);
+        if (astVarp->ignoreSchedWrite()) return true;
+        return false;
+    }
+
+    void updateGen(AstNode* nodep) {
+        nodep->foreach([&](AstNodeVarRef* refp) {
```

Comment:
```suggestion
        nodep->foreach([&](AstNodeVarRef* refp) {
             if (m_abort) return;
```

Short circuit on aborts?